File: python/Combination Sum III.py
class Solution(object):
    def combinationSum3(self, k, n):
        stack, ans = [([], 0, n)], []
        
        while stack:
            tmp, index, target = stack.pop()
            if len(tmp) >= k:
                continue
            for i in range(index + 1, 10):
                if target == i and len(tmp) == k - 1:
                    ans.append(tmp + [i])
                elif target > i:
                    stack.append((tmp + [i], i, target - i))
        
        return ans

File: cpp/Combination Sum III.cpp
class Solution {
public:
    vector< vector<int> > answer;
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<int> tmp;
        for (int i = 1; i <= 9; i++) {
            tmp.push_back(i);
            helper(i, k - 1, n - i, tmp);
            tmp.pop_back();
        }
        return answer;
    }
    
    void helper(int cur, int count, int target, vector<int> &v) {
        if (count < 0) {
            return;
        }
        if (count == 0) {
            if (target == 0) {
                answer.push_back(v);
            }
            return;
        }
        for (int i = cur + 1; i <= 9; i++) {
            v.push_back(i);
            helper(i, count - 1, target - i, v);
            v.pop_back();
        }
    }
};
